src/display/graphics.c

```diff
@@ -0,0 +1,387 @@
+	char* out;
+	vasprintf(&out, fmt, args);
+	va_list empty;
+	switch(txt_fmt){
+		case 0:
+		case 1:
+		{
+            vexDisplayVString(line, out, empty);
+            break;
+        }
+		case 2:
+		{
+            vexDisplayVBigString(line, out, empty);
+            break;
+        }
+        case 3:
+		{
+            vexDisplayVCenteredString(line, out, empty);
+            break;
+        }
+        case 4:
+        {
+            vexDisplayVBigCenteredString(line, out, empty);
+            break;
+        }
+        default:
+        {
+            vexDisplayVString(line, out, empty);
+            break;
+        }
+    }
+	mutex_give(_display_mutex);
+}
+
+void display_vprintf_at(uint8_t txt_fmt, const int16_t x, const int16_t y, const char* fmt, va_list args){
+	mutex_take(_display_mutex, TIMEOUT_MAX);
+	char* out;
+	vasprintf(&out, fmt, args);
+	va_list empty;
+	switch(txt_fmt){
+		case 0:
+		{
+            vexDisplayVSmallStringAt(x, y, out, empty);
+            break;
+        }
+		case 1:
+		case 3:
+		{
+            vexDisplayVStringAt(x, y, out, empty);
+            break;
+        }
+		case 2:
+		case 4:
+		{
+            vexDisplayVBigStringAt(x, y, out, empty);
+            break;
+        }
+        default:
+        {
+            vexDisplayVStringAt(x, y, out, empty);
+            break;
+        }
+    }
+	mutex_give(_display_mutex);
+}
+
+void display_opacity_vprintf( int32_t xpos, int32_t ypos, uint32_t opacity, const char *format, va_list args) {
+	mutex_take(_display_mutex, TIMEOUT_MAX);
+	char* out;
+	vasprintf(&out, format, args);
+	va_list empty;
+	vexDisplayVPrintf(xpos, ypos, opacity, out, empty);
+	mutex_give(_display_mutex);
+}
+
+int16_t touch_last_x() {
+	static V5_TouchStatus v5_touch_status;
+	mutex_take(_display_mutex, TIMEOUT_MAX);
+	vexTouchDataGet(&v5_touch_status);
+	mutex_give(_display_mutex);
+	return v5_touch_status.lastXpos;
+}
+
+int16_t touch_last_y() {
+	static V5_TouchStatus v5_touch_status;
+	mutex_take(_display_mutex, TIMEOUT_MAX);
+	vexTouchDataGet(&v5_touch_status);
+	mutex_give(_display_mutex);
+	return v5_touch_status.lastYpos;
+}
+
+touch_event_e_t touch_last_event() {
+	static V5_TouchStatus v5_touch_status;
+	mutex_take(_display_mutex, TIMEOUT_MAX);
+	vexTouchDataGet(&v5_touch_status);
+	mutex_give(_display_mutex);
+	return v5_touch_status.lastEvent;
+}
+
+static linked_list_s_t* _touch_event_release_handler_list = NULL;
+static linked_list_s_t* _touch_event_press_handler_list = NULL;
+static linked_list_s_t* _touch_event_press_auto_handler_list = NULL;
+
+void _set_up_touch_callback_storage() {
+	_touch_event_release_handler_list = linked_list_init();
+	_touch_event_press_handler_list = linked_list_init();
+	_touch_event_press_auto_handler_list = linked_list_init();
+}
+
+void register_touch_callback(touch_event_cb_fn_t cb, touch_event_e_t event_type) {
+	switch (event_type) {
+	case E_TOUCH_EVENT_RELEASE:
+		linked_list_prepend_func(_touch_event_release_handler_list, (generic_fn_t)cb);
```

Comment:
What's the thread safety of this and related APIs?

Reply:
I haven't done any work to make this thread safe yet... I didn't expose this functionality yet. Potentially will do soon or in a later update.

Reply:
Alright, I'll add a mutex lock here.